### PR TITLE
Parse Private Names

### PR DESCRIPTION
**Pull request by mheiber**
## Description of Changes

Parser and Scanner:
- Add new SyntaxKind: SyntaxKind.PrivateName
- Edit baseline reference files to use `¬` (hook) instead of `#` (hash) to test parser error recovery.
- Test for correct parsing
- "Property _ does not exist on type _" error should still happen with private names even if the class has an index signature (new test case)

## Steps to Test

- Build the compiler and services (`npm run build:compiler`)
- Tests should pass. Please review the new test files in the `privateNames` directory carefully, the rest are minor tweaks that I hope don't matter.
- Check the AST. Assuming you have a file in the right place with a private name declaration and access, the AST should have private name nodes in both places. 

`node -e "ts = require('./built/local/tsserver');visit = node => {console.log(node); ts.forEachChild(node, visit)};visit(ts.createSourceFile('f', fs.readFileSync('../tssample/src/1.ts', 'utf8')))`

> The binder and checker work for private names in this PR is very primitive, just a start

### CODE REVIEW

File: src/compiler/visitor.ts [hunk 1]
```diff
@@ -221,6 +221,8 @@ namespace ts {
 
             case SyntaxKind.Identifier:
                 return updateIdentifier(<Identifier>node, nodesVisitor((<Identifier>node).typeArguments, visitor, isTypeNodeOrTypeParameterDeclaration));
+            case SyntaxKind.PrivateName:
+                return updatePrivateName(node as PrivateName);
```
Comment (joeywatts) [inline src/compiler/visitor.ts]: If no modification is needed here, we can probably remove this case, right?

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -1386,6 +1389,17 @@ namespace ts {
             return finishNode(node);
         }
 
+        function createPrivateName(): PrivateName {
```
Comment (joeywatts) [inline src/compiler/parser.ts]: Can we just call this `parsePrivateName` and get rid of the function below?

File: src/compiler/types.ts [hunk 3]
```diff
@@ -741,6 +742,13 @@ namespace ts {
         expression: Expression;
     }
 
+    export interface PrivateName extends Declaration {
```
Comment (RyanCavanaugh) [inline src/compiler/types.ts]: Should extend `Node` not `Declaration`

File: src/compiler/types.ts [hunk 4]
```diff
@@ -1176,7 +1184,7 @@ namespace ts {
 
     export interface StringLiteral extends LiteralExpression {
         kind: SyntaxKind.StringLiteral;
-        /* @internal */ textSourceNode?: Identifier | StringLiteralLike | NumericLiteral; // Allows a StringLiteral to get its text from another node (used by transforms).
+        /* @internal */ textSourceNode?: Identifier | PrivateName | StringLiteralLike | NumericLiteral; // Allows a StringLiteral to get its text from another node (used by transforms).
```
Comment (RyanCavanaugh) [inline src/compiler/types.ts]: I don't think this case should occur (?)

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -2641,8 +2653,8 @@ namespace ts {
         return node.kind === SyntaxKind.Identifier && (<Identifier>node).escapedText === "Symbol";
     }
 
-    export function isPushOrUnshiftIdentifier(node: Identifier) {
-        return node.escapedText === "push" || node.escapedText === "unshift";
+    export function isPushOrUnshiftIdentifier(node: Identifier | PrivateName) {
```
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: Shouldn't be legal to invoke this with a `PrivateName`

File: src/compiler/utilities.ts [hunk 6]
```diff
@@ -770,12 +771,21 @@ namespace ts {
 
     export function entityNameToString(name: EntityNameOrEntityNameExpression): string {
         switch (name.kind) {
+            // fall through
```
Comment (RyanCavanaugh) [inline src/compiler/utilities.ts]: Seems to be have left in unintentionally

**Comment by RyanCavanaugh** [comment]
Possible bug (haven't tried it but there should be a testcase):
```ts
class foo {
    [k: string]: any;
    b() {
        // Should error 
        this.#x = 10;
    }
}
```

File: src/compiler/types.ts (hunk 3 above)
Comment (mheiber) [reply src/compiler/types.ts]: thanks @RyanCavanaugh ! I'll check this out.

File: src/compiler/types.ts (hunk 4 above)
Comment (mheiber) [reply src/compiler/types.ts]: @RyanCavanaugh thanks for noticing this, I wasn't sure what to do here. In factory.ts there is a helper function for creating literals from nodes, it seemed that we would want this to work from private names as well:

https://github.com/bloomberg/TypeScript/pull/4/files#diff-9c914fcc05c72a72db62c0a023b15f3fR110

File: src/compiler/types.ts (hunk 3 above)
Comment (RyanCavanaugh) [reply src/compiler/types.ts]: Ah, I was mistaken on this one. `extends Declaration` is fine

**Comment by mheiber** [comment]
@RyanCavanaugh you were right that this doesn't error (but should):

```ts
class foo {
    [k: string]: any;
    b() {
        // Should error 
        this.#x = 10;
    }
}
```

1) Do you have guidance on what is causing this problem? It seems that we'll need to encode the information that private names are not accessible via indexing with a string.
2) Since this PR is just for getting private names to parse and the checker work is [started in a follow-on PR](https://github.com/bloomberg/TypeScript/pull/5), what do you think of deferring it?

**Comment by RyanCavanaugh** [comment]
@mheiber you'd probably just want to add a baseline test, which doesn't directly encode any "*this* should happen" information, so I think it would make sense to just add the testcase and then let the next PR update its baseline outputs

**Comment by RyanCavanaugh** [comment]
As for behavior, in `checker.ts` see `checkPropertyAccessExpressionOrQualifiedName`:
```ts
            const prop = getPropertyOfType(apparentType, right.escapedText);
            if (isIdentifier(left) && parentSymbol && !(prop && isConstEnumOrConstEnumOnlyModule(prop))) {
                markAliasReferenced(parentSymbol, node);
            }
            if (!prop) {
HERE        ^^^^^^^^^^^^
                const indexInfo = getIndexInfoOfType(apparentType, IndexKind.String);
                if (!(indexInfo && indexInfo.type)) {
                    if (right.escapedText && !checkAndReportErrorForExtendingInterface(node)) {
                        reportNonexistentProperty(right, leftType.flags & TypeFlags.TypeParameter && (leftType as TypeParameter).isThisType ? apparentType : leftType);
                    }
                    return errorType;
                }
                if (indexInfo.isReadonly && (isAssignmentTarget(node) || isDeleteTarget(node))) {
                    error(node, Diagnostics.Index_signature_in_type_0_only_permits_reading, typeToString(apparentType));
                }
                propType = indexInfo.type;
            }
```

Here, we get `prop` by looking up the dotted name. If we don't find anything, then there's no declared property with this name. When that happens, we check for an index signature, and if there's a suitable string index signature, we act as if there's a property with that name of the type of the index signature.

The diff should be approximately as simple as
```ts
if (!prop && /* the property name isn't a private name */) {
          +++++++++++++++++++++++++++++++++++++++++++++++
```

File: src/compiler/types.ts (hunk 4 above)
Comment (mheiber) [reply src/compiler/types.ts]: ping @RyanCavanaugh

Comment (RyanCavanaugh) [reply src/compiler/types.ts]: `createLiteralFromNode` shouldn't accept a `PrivateName` node - unlike property names (whose names are meaningful in isolation), private names can't be used anywhere a literal is needed.

Any place where `createLiteral` is being called today with a node that is sourced from a property name, we need to add logic to do the right thing for `PrivateName`s. This should just be mostly `assert`s to make sure we don't e.g. try to rewrite `x.#foo` with `x["#foo"]` (something we currently treat as being always OK).

Comment (mheiber) [reply src/compiler/types.ts]: I incorporated this feedback and [disallowed creating literals from private names](https://github.com/bloomberg/TypeScript/pull/4/commits/de03075ebaac584628cf49eb58d88b442b0ea2ff).

I think the code does what we want now, with these caveats:
- [this type alias smells a little](https://github.com/bloomberg/TypeScript/pull/4/commits/de03075ebaac584628cf49eb58d88b442b0ea2ff#diff-9c914fcc05c72a72db62c0a023b15f3fR3585). I used it to avoid making big changes to the type hierarchy in types.d.ts.
- I also had to stick a [TODO in the transformer](https://github.com/bloomberg/TypeScript/pull/4/commits/de03075ebaac584628cf49eb58d88b442b0ea2ff#diff-db4815b580d3d6fd21c975e36c8f81d1R1645).

Comment (mheiber) [reply src/compiler/types.ts]: ping @RyanCavanaugh would you be able to take another look?

File: src/compiler/transformers/destructuring.ts [hunk 7]
```diff
@@ -535,7 +535,8 @@ namespace ts {
         const propertyNames: Expression[] = [];
         let computedTempVariableOffset = 0;
         for (let i = 0; i < elements.length - 1; i++) {
-            const propertyName = getPropertyNameOfBindingOrAssignmentElement(elements[i]);
+            // can safely assume property name is not a PrivateName because PrivateNames are nto allowed in object literals
```
Comment (joeywatts) [inline src/compiler/transformers/destructuring.ts]: typo, "PrivateNames are **not** allowed in object literals"

File: src/compiler/utilities.ts [hunk 8]
```diff
@@ -2642,7 +2653,7 @@ namespace ts {
     }
 
     export function isPushOrUnshiftIdentifier(node: Identifier) {
-        return node.escapedText === "push" || node.escapedText === "unshift";
+        return isIdentifier(node) && node.escapedText === "push" || node.escapedText === "unshift";
```
Comment (Neuroboy23) [inline src/compiler/utilities.ts]: Isn't `isIdentifier(node)` always `true` here?

Comment (mheiber) [reply src/compiler/utilities.ts]: nice catch. True, the node will always be an identifier, fixed

File: src/compiler/factory.ts [hunk 9]
```diff
@@ -3581,11 +3581,17 @@ namespace ts {
         }
     }
 
+    /**
+     * accessor declaration that can be converted to an expression (`name` field cannot be a `PrivateName`)
+     */
+    type ExpressionableAccessorDeclaration = AccessorDeclaration & {name: Exclude<PropertyName, PrivateName>};
```
Comment (joeywatts) [inline src/compiler/factory.ts]: Might be more consistent to define like this:

```typescript
interface ExpressionableAccessorDeclaration extends AccessorDeclaration {
    name: Exclude<PropertyName, PrivateName>;
}
```

Comment (mheiber) [reply src/compiler/factory.ts]: Yours looks better to me, too, but unfortunately AccessorDeclaration is a type, not an interface, and so cannot be extended. So I'd end up with a type intersection anyway.

File: src/compiler/binder.ts [hunk 10]
```diff
@@ -266,6 +266,9 @@ namespace ts {
                     Debug.assert(isWellKnownSymbolSyntactically(nameExpression));
                     return getPropertyNameForKnownSymbolName(idText((<PropertyAccessExpression>nameExpression).name));
                 }
+                if (isPrivateName(node)) {
+                    return nodePosToString(node) as __String;
+                }
```
Comment (mheiber) [inline src/compiler/binder.ts]: ignore this: it's a mistake and to be removed in https://github.com/bloomberg/TypeScript/pull/5. Actually has no effect, since there is no way the `node` can be a private name here anyway.

File: src/compiler/checker.ts [hunk 11]
```diff
@@ -21876,6 +21878,9 @@ namespace ts {
             checkGrammarDecoratorsAndModifiers(node);
 
             checkVariableLikeDeclaration(node);
+            if (node.name && isIdentifier(node.name) && node.name.originalKeywordKind === SyntaxKind.PrivateName) {
+                error(node, Diagnostics.Private_names_cannot_be_used_as_parameters);
```
Comment (mheiber) [inline src/compiler/checker.ts]: we plan to undo this change is in https://github.com/bloomberg/TypeScript/pull/5

File: src/compiler/checker.ts [hunk 12]
```diff
@@ -29288,6 +29293,10 @@ namespace ts {
                 checkESModuleMarker(node.name);
             }
 
+            if (isIdentifier(node.name) && node.name.originalKeywordKind === SyntaxKind.PrivateName) {
+                return grammarErrorOnNode(node.name, Diagnostics.Private_names_are_not_allowed_in_variable_declarations);
+            }
```
Comment (mheiber) [inline src/compiler/checker.ts]: we plan to undo this change in https://github.com/bloomberg/TypeScript/pull/5

File: src/compiler/transformers/destructuring.ts [hunk 13]
```diff
@@ -535,7 +535,8 @@ namespace ts {
         const propertyNames: Expression[] = [];
         let computedTempVariableOffset = 0;
         for (let i = 0; i < elements.length - 1; i++) {
-            const propertyName = getPropertyNameOfBindingOrAssignmentElement(elements[i]);
+            // can safely assume property name is not a PrivateName because PrivateNames are not allowed in object literals
+            const propertyName = getPropertyNameOfBindingOrAssignmentElement(elements[i]) as Exclude<PropertyName, PrivateName>;
```
Comment (mheiber) [inline src/compiler/transformers/destructuring.ts]: should I add a Debug.assert here? Don't want this to be a bug vector if private names are eventually allowed in object literals
